### ISSUE TITLE
Exemption for for...of unused variable.

### ISSUE THREAD

**Issue by dimaqq** [eslint/eslint#12287]
In similar vein to `for...in` https://github.com/eslint/eslint/issues/2342
Should there be similar exemption for [async] `for...of`?

**What rule do you want to change?**
`no-unused-vars`

**Does this change cause the rule to produce more or fewer warnings?**
fewer warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
I don't know

**Please provide some example code that this change will affect:**

```js
async function forever(x) {
  for await (const unused of x) {
    void unused;
  }
}

forever(monitor_network());
```

**What does the rule currently do for this code?**
`  Line 42:  'unused' is defined but never used  no-unused-vars`

**What will the rule do after it's changed?**
Nothing

**Are you willing to submit a pull request to implement this change?**
This seems tricky...

**Comment by mqklin** [comment]
This is broken for me after eslint update to v6 as well:
```
for (const x of y) { // x is defined but never used (no-unused-vars)
  console.log(x);
}
```

**Comment by kaicataldo** [comment]
This looks like a duplicate of https://github.com/eslint/eslint/issues/12117. Can you see if the solutions there fix your issue?

**Comment by kaicataldo** [comment]
Note that this issue doesn't show in the [ESLint Demo ](https://eslint.org/demo#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZm9yZXZlcih4KSB7XG4gICAgZm9yIGF3YWl0IChjb25zdCB1bnVzZWQgb2YgeCkge1xuICAgICAgICB2b2lkIHVudXNlZDtcbiAgICB9XG59XG5cbmZvcmV2ZXIobW9uaXRvcl9uZXR3b3JrKCkpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6MiwiaW5kZW50IjoyfSwiZW52Ijp7fX19).

Are you using babel-eslint?

**Comment by mqklin** [comment]
Yes, I use babel-eslint. Yes, updating babel-eslint fixed the issue, thanks a lot!

**Comment by g-plane** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the [ESLint Gitter](https://gitter.im/eslint/eslint) if you have any other issues!

**Comment by platinumazure** [comment]
@g-plane I could be missing something, but... I'm not sure this should be closed. The original post was a rule enhancement request for "for await of", but then someone else came and asked about "for of" and we noted the babel-eslint/eslint-scope issue.

Should we still evaluate the original rule request? If not, why not?

**Comment by g-plane** [comment]
Oh, it seems reasonable. Sorry.

**Comment by kaicataldo** [comment]
I'm not able to reproduce with our [demo](https://eslint.org/demo#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZm9yZXZlcih4KSB7XG4gICAgZm9yIGF3YWl0IChjb25zdCB1bnVzZWQgb2YgeCkge1xuICAgICAgICB2b2lkIHVudXNlZDtcbiAgICB9XG59XG5cbmZvcmV2ZXIobW9uaXRvcl9uZXR3b3JrKCkpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6MiwiaW5kZW50IjoyfSwiZW52Ijp7fX19).

@dimaqq I see you left a response on my comment above - did this solve your problem?

**Comment by dimaqq** [comment]
I can only test it on Monday, sorry.
I think that with `void` trick or ought to work; and without it I’m not sure...

**Comment by mysticatea** [comment]
I guess you can use [varsIgnorePattern](https://eslint.org/docs/rules/no-unused-vars#varsignorepattern) option to ignore unused variables by naming convention.

For example, [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtlcnJvciwgeyB2YXJzSWdub3JlUGF0dGVybjogXCJeX1wiIH1dICovXG5cbmFzeW5jIGZ1bmN0aW9uIGZvcmV2ZXIoeCkge1xuICBmb3IgYXdhaXQgKGNvbnN0IF91bnVzZWQgb2YgeCkge1xuICB9XG59XG5cbmZvcmV2ZXIobW9uaXRvcl9uZXR3b3JrKCkpO1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**Comment by dimaqq** [comment]
There's something wrong with the online demo, it says `Lint-free` even for code like this:
```

**Comment by kaicataldo** [comment]
None of the rules are on in the link you provided. When I turn on `no-unused-vars`, it’s displaying an error as expected.

**Comment by dimaqq** [comment]
I'm so sorry for the confusion.

When I read #2342 I was left with the impression that the following is allowed:

```js
for (unused in something) {}
```

But of course that's only because `unused` could be anything here, e.g. a global, and other rules kick in now: `no-undef`, `prefer-const`.

It didn't think that the following is semantically different wrt. the unused variable rule and treated both cases the same:

```js
for (const unused in something) {}
```

Weirdly there's test case for the above, as if that's valid code: https://github.com/eslint/eslint/blob/76fb571a1c15b040e42272435eb0d023cdcb031e/tests/lib/rules/no-unused-vars.js#L159
Yet, if I "enable all rules" in the eslint demo, that very example triggers `no-unused-vars` 😕 

In any case `void unused;` works and keeps eslint silent.

Also, as far I can tell `for await (const x of y)` is treated exactly same as `for (const x of y)`.

I guess I'm OK then :)

**Comment by kaicataldo** [comment]
Thanks for following up! I'm going to go ahead and close this issue then.
